### PR TITLE
[tune] Function API checkpointing

### PR DESCRIPTION
**Pull request by richardliaw**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tune/session.py [hunk 1]
```diff
@@ -86,7 +63,72 @@ def run_me(config):
...
+
+    .. code-block:: python
+
+        import time
+        from ray import tune
+
+        def run_me(config, checkpoint=0):
+            for iter in range(checkpoint, 100):
+                time.sleep(1)
+                tune.save_checkpoint(iter)
```
Comment (richardliaw) [inline python/ray/tune/session.py:103]: fix

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
It looks good, but please make sure the RLlib tests all pass. It is currently causing some checkpoint related test failures

**Comment by richardliaw** [comment]
cc @krfricke

File: python/ray/tune/tests/test_cluster.py [hunk 2]
```diff
@@ -394,41 +394,49 @@ def test_migration_checkpoint_removal(start_connected_emptyhead_cluster,
...
-            mock_find_dir.side_effect = mock_find_dir_fn
-
             # Test recovery of trial that has been checkpointed
             t1 = Trial(trainable_id, **kwargs)
             runner.add_trial(t1)
 
             # Start trial, process result (x2), process save
             for _ in range(4):
                 runner.step()
+            print(t1.last_result)
```
Comment (richardliaw) [inline python/ray/tune/tests/test_cluster.py:433]: remove

File: python/ray/tune/function_runner.py [hunk 3]
```diff
@@ -264,13 +351,35 @@ def _report_thread_runner_error(self, block=False):
...
+                logger.warning(
+                    "Function checkpointing is disabled. This may result in "
+                    "unexpected behavior when using checkpointing features or "
+                    "certain schedulers. To enable, set the train function "
+                    "arguments to be `func(config, checkpoint)`.")
                 output = train_func(config)
+            elif use_checkpoint:
+                output = train_func(config, checkpoint=checkpoint)
             else:
                 output = train_func(config, reporter)
```
Comment (krfricke) [inline python/ray/tune/function_runner.py:384]:
```suggestion
                output = train_func(config, reporter=reporter)
```

File: python/ray/tune/tune.py [hunk 4]
```diff
@@ -154,8 +154,10 @@ def run(run_or_experiment,
             syncing to driver is disabled.
         checkpoint_freq (int): How many training iterations between
             checkpoints. A value of 0 (default) disables checkpointing.
+            This has no effect when using the Functional Training API.
         checkpoint_at_end (bool): Whether to checkpoint at the end of the
             experiment regardless of the checkpoint_freq. Default is False.
+            This has no effect when using the Functional API.
```
Comment (krfricke) [inline python/ray/tune/tune.py:160]:
```suggestion
            This has no effect when using the Functional Training API.
```

**Review by krfricke — APPROVED** [review]

**Pull request merged**